fw/http_frame.c

```diff
@@ -1894,21 +1894,25 @@ tfw_h2_calc_frame_length(TfwH2Ctx *ctx, TfwStream *stream, TfwFrameType type,
 static inline char
 tf2_h2_calc_frame_flags(TfwStream *stream, TfwFrameType type)
 {
-	switch (type) {
-	case HTTP2_HEADERS:
-		return stream->xmit.h_len ?
-			(stream->xmit.b_len ? 0 : HTTP2_F_END_STREAM) :
-			(stream->xmit.b_len ? HTTP2_F_END_HEADERS :
-			 HTTP2_F_END_HEADERS | HTTP2_F_END_STREAM);
-	case HTTP2_CONTINUATION:
-		return stream->xmit.h_len ? 0 : HTTP2_F_END_HEADERS;
-	case HTTP2_DATA:
-		return stream->xmit.b_len ? 0 : HTTP2_F_END_STREAM;
-	default:
-		BUG();
-	};
+	char flags = 0;
 
-	return 0;
+	if (stream->xmit.h_len && !stream->xmit.b_len
```

Comment:
According to RFC 9113 8.1:

Trailer fields are carried in a field block that also terminates the stream. That is, trailer fields comprise a sequence starting with a [HEADERS](https://www.rfc-editor.org/rfc/rfc9113.html#HEADERS) frame, followed by zero or more [CONTINUATION](https://www.rfc-editor.org/rfc/rfc9113.html#CONTINUATION) frames, where the [HEADERS](https://www.rfc-editor.org/rfc/rfc9113.html#HEADERS) frame bears an END_STREAM flag. Trailers MUST NOT include [pseudo-header fields](https://www.rfc-editor.org/rfc/rfc9113.html#PseudoHeaderFields) ([Section 8.3](https://www.rfc-editor.org/rfc/rfc9113.html#PseudoHeaderFields)). An endpoint that receives pseudo-header fields in trailers MUST treat the request or response as [malformed](https://www.rfc-editor.org/rfc/rfc9113.html#malformed) ([Section 8.1.1](https://www.rfc-editor.org/rfc/rfc9113.html#malformed)).

So if trailer exists we should set HTTP2_F_END_STREAM after sending HEADERS frame for trailers.

Reply:
This comment is not resolved. If trailers are present we should not set HTTP2_F_END_STREAM here according RFC

Reply:
What do you mean? Please see this [commit](https://github.com/tempesta-tech/tempesta/pull/2173/commits/372dbb0e17b1172a8e9e22af2d01008de52be96b).

If there are trailers, the body would not be empty. So the code you refer to cannot be true: `!stream->xmit.b_len`.

Reply:
done: https://github.com/tempesta-tech/tempesta/pull/2173/commits/7c4107e41c246a346d7edb965a039e6d0e247285